mindmeld/active_learning/alp.py

```diff
@@ -0,0 +1,234 @@
+            labeled_logs_pattern (str): Pattern to obtain logs already labeled in a MindMeld app
+            unlabeled_logs_path (str): Path a logs text file with unlabeled queries
+            output_folder (str):
+        """
+        self.app_path = app_path
+        self.train_pattern = train_pattern
+        self.test_pattern = test_pattern
+        self.load = load
+        self.save = save
+        self.train_seed_pct = train_seed_pct
+        self.n_classifiers = n_classifiers
+        self.n_epochs = n_epochs
+        self.batch_size = batch_size
+        self.training_level = training_level
+        self.training_strategies = training_strategies
+        self.log_selection_strategy = log_selection_strategy
+        self.save_sampled_queries = save_sampled_queries
+        self.early_stopping_window = early_stopping_window
+        self.log_usage_pct = log_usage_pct
+        self.labeled_logs_pattern = labeled_logs_pattern
+        self.unlabeled_logs_path = unlabeled_logs_path
+
+        self.output_manager = self.get_output_manager(output_folder)
+        self.mindmeld_classifier = self.get_classifier()
+        self.init_data_bucket = None
+        self.data_bucket = None
+
+    def get_output_manager(self, output_folder):
+        return OutputManager(
+            active_learning_params=deepcopy(self.__dict__), output_folder=output_folder
+        )
+
+    def get_classifier(self):
+        return MindMeldClassifier(
+            app_path=self.app_path,
+            training_level=self.training_level,
+            n_classifiers=self.n_classifiers,
+        )
+
+    def train(self):
+        """Loads the initial data bucket and then trains on every strategy."""
+        logger.info("Creating Training Data Bucket")
+        self.output_manager.create_experiment_folder(self.training_strategies)
+        self.init_data_bucket = DataBucketFactory.get_data_bucket_for_training(
+            self.app_path,
+            self.load,
+            self.save,
+            self.train_pattern,
+            self.test_pattern,
+            self.train_seed_pct,
+        )
+        logger.info("Starting Training")
+        for strategy in self.training_strategies:
+            self._train_strategy(strategy)
+
+    def _train_strategy(
+        self,
+        strategy: str,
+        selection_mode: bool = None,
+    ):
+        """Helper function to traing a single strategy.
+
+        Args:
+            strategy (str): Single strategy to train
+            selection_mode (bool): If in selection mode, accuracies will not be recorded
+                and the run will terminate after the first iteration
+        """
+        early_stop = False
+        for epoch in range(self.n_epochs):
+            del self.data_bucket
+            self.data_bucket = deepcopy(self.init_data_bucket)
+            num_iterations = math.ceil(
+                len(self.data_bucket.unsampled_queries) / self.batch_size
```

Comment:
I think this computation can be outside the for loop, unless self.data_bucket is mutating at each iteration